File: C/my_c/step0_repl.c
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <readline/readline.h>
#include <readline/history.h>

char * READ(char prompt[])
{
  char *line;
  line = readline(prompt);
  if(!line) return NULL;
  add_history(line);
  return line;
}

char *EVAL(char *ast,void *env)
{
  return ast;
}

char *PRINT(char *exp)
{
  return exp;
}

int main()
{
  char *ast,*exp;
  char prompt[100];

  snprintf(prompt,sizeof(prompt),"user> ");

  for(;;)
    {
      ast = READ(prompt);
      if(!ast) return 0;
      exp = EVAL(ast,NULL);
      puts(PRINT(exp));

      free(ast);
    }

}

File: nodejs/mal/step0_repl.js

var readline = require('./node_readline');
var printer = require('./printer');

function READ(str){
	return str;
};

function EVAL(ast, env){
	return ast;
};

function PRINT(str){
	return str;
};

var rep =  function(str) { return PRINT(EVAL(READ(str),{})); };

// main-loop

while(true){
	var line = readline.readline("user> ");
	if(line === null) {break;}
	try{
		if(line) {printer.println(rep(line));}
	}catch(err){
		if(err.stack) {printer.println(err.stack);}
		else {printer.println(err);}
	}
}
